west/cmd/git.py

```diff
@@ -0,0 +1,255 @@
+
+'''west git commands'''
+
+import argparse
+import collections
+import os
+import subprocess
+import sys
+
+import pykwalify.core
+import yaml
+
+from . import WestCommand
+from .. import log
+
+
+class Sync(WestCommand):
+    def __init__(self):
+        super().__init__(
+            'sync',
+            'Clone/update all Git repositories specified in the manifest file')
+
+    def do_add_parser(self, parser_adder):
+        return _add_common_git_flags(parser_adder, self)
+
+    def do_run(self, args, user_args):
+        projects = _all_projects(args)
+
+        # TODO: Error checking (if _git(...).returncode != 0: ...)
+
+        for project in projects:
+            if not os.path.exists(project.path):
+                _git_top(project, 'clone -b (branch) (url)/(name) (path)')
+            else:
+                # Fetch first to make sure the project's branch exists. It
+                # might not if 'git clone' was aborted, for example.
+                _git(project, 'fetch')
+
+                # TODO: The local branch might not exist if the manifest was
+                # updated. Maybe local branches could be stored in a separate
+                # namespace...
+
+                # Check out the main branch and update it
+                _git(project, 'checkout (branch)')
+                _git(project, 'rebase FETCH_HEAD')
+
+                # Switch back to previous branch the user was on
+                _git(project, 'checkout -')
+
+
+class Diff(WestCommand):
+    def __init__(self):
+        super().__init__(
+            'diff',
+            "Run 'git diff' for each project. Extra arguments are passed "
+            "as-is to 'git diff'.",
+            accepts_unknown_args=True)
+
+    def do_add_parser(self, parser_adder):
+        return _add_common_git_flags(parser_adder, self)
+
+    def do_run(self, args, user_args):
+        for project in _all_projects(args):
+            if _check_repo(project):
+                # Use paths that are relative to the base directory to show
+                # which repo any changes are in
+                _git(project, 'diff --src-prefix=(path)/ --dst-prefix=(path)/',
+                     extra_args=user_args)
+
+
+class Status(WestCommand):
+    def __init__(self):
+        super().__init__(
+            'status',
+            "Run 'git status' for each project. Extra arguments are passed "
+            "as-is to 'git status'.",
+            accepts_unknown_args=True)
+
+    def do_add_parser(self, parser_adder):
+        return _add_common_git_flags(parser_adder, self)
+
+    def do_run(self, args, user_args):
+        for project in _all_projects(args):
+            if _check_repo(project):
+                _git(project, 'fetch')
+
+                print("=== 'git status' for {} (in {}) ==="
+                      .format(project.name, project.path))
+                _git(project, 'status', extra_args=user_args)
+
+
+def _add_common_git_flags(parser_adder, command):
```

Comment:
Ah, also: it would be nice to be able to take a `nargs='*'` argument specifying the list of projects to synchronize, diff against, or get status for.

Supporting either the name of the project in the manifest or its locally checked out path would be most convenient.

Reply:
Yeah, would be nice to have later